### ISSUE TITLE
`Task.cancel` has a bad behavior

### ISSUE THREAD

**Issue by roy2220** [MagicStack/uvloop#166]
I'm a uvloop fans, nowadays I found a bad behavior of `Task.cancel`:

```python
import asyncio

class Foo:
    def start(self):
        self.task = loop.create_task(self.run())

    def stop(self):
        self.task.cancel()
        self.task = None

    async def run(self):
        print("--- run begin---")
        try:
            await loop.create_future()
        except asyncio.CancelledError:
            pass
        print("--- run end ---")

    def __del__(self):
        if self.task and not loop.is_closed():
            self.task.cancel()

async def test():
    cur_task = asyncio.Task.current_task()

    async def get_foo():
        foo = Foo()
        foo.start()
        print("got foo!")
        return foo

    def cb():
        cur_task.cancel()

    getting_foo = loop.create_task(get_foo())
    loop.call_soon(cb)

    try:
        f = await getting_foo
    except asyncio.CancelledError:
        pass
    else:
        f.stop()

    del getting_foo
    print("sleep 1 sec wait for '--- run end ---' ...")
    await asyncio.sleep(1)

loop = asyncio.get_event_loop()
loop.run_until_complete(test())
loop.close()
```

print "--- run end ---" is expected but NEVER!
The `Foo` object seems "leaked".

I knew it is not a bug of uvloop, it just is a bad design of asyncio.?
How do you regard this behavior? It would be great if you could give me some advices.

**Comment by 1st1** [comment]
Looks like you're cancelling the `test()` task with `loop.call_soon(cb)` and the cancellation happens before `await getting_foo` and `Foo.run()` "connect" with each other.  If you change `loop.call_soon(cb)` to `loop.call_later(0.1, cb)` the script works as expected.

This might be a bug with `asyncio.Task`, but your code snippet is a bit hard to parse and understand.  I suggest you to spend some time and simplify your code so that it's more obvious what's going on here.  Then I suggest you to submit a bug to bugs.python.org!

**Comment by roy2220** [comment]
@1st1 Thanks for you time and advice! That is a subtle situation, I spend a whole day to figure out the behavior of `Task.cancel`, finally I made a patch to `Task.cancel` for more rational behavior. Here is the [commit](https://github.com/roy2220/cpython/commit/81b92f5b21ea9d26a32456c044b074b2a160ad49) (Actually I patched the `_asyncio` C module, but here is python for readability). In conclusion, I regard it as a bad design of `Task.cancel` behavior. I want to submit a bug to bugs.python.org, but this issue is subtle and I'm not good at English, it's somewhat difficult (for me) to describe the situation.
